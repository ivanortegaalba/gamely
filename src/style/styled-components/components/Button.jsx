import styled from 'styled-components'
import { key, marginFromPropsMixing, palette, px2rems } from '../tools/functions'
import PropTypes from 'prop-types'
import { colors } from '../../styled-components/settings/colors'
import { spacing } from '../../styled-components/settings/spacing'

const radius = px2rems(2)

const padding = {
  s: `${px2rems(7)}`,
  m: `${px2rems(11)} ${px2rems(16)}`,
  l: `${px2rems(11)} ${px2rems(23)}`
}

const size = () => {
  return (props) => padding[props.size]
}

export const Button = styled.button`
  background-color: ${(props) => Boolean(props.reverse)
    ? key('palette.white')
    : palette(1)};
  border: 1px solid ${palette(1)};
  border-radius: ${radius};
  color: ${(props) => Boolean(props.reverse)
    ? palette(2)
    : key('palette.white') };
  cursor: pointer;
  font-variant: all-small-caps;
  font-weight: 600;
  padding: ${size};
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;

  &:hover {
  background-color: ${key('palette.white')};
  border-color:${palette(0)};
  color: ${palette(0)}};
  }

  &:active,
  &:focus {
    border-color: ${palette(2)};
  }

  &[disabled],
  &[disabled]:hover {
    background-color: ${key('palette.white')};
    border-color:${palette(2)};
    color: ${palette(2)};
    cursor: inherit;
  }
  
  ${marginFromPropsMixing}
`

Button.propTypes = {
  palette: PropTypes.oneOf(Object.keys(colors)),
  padding: PropTypes.oneOf(Object.keys(spacing)),
  margin: PropTypes.string,
  reverse: PropTypes.bool
}
Button.defaultProps = {
  palette: 'default',
  size: 'm',
  margin: '0',
  reverse: false
}
