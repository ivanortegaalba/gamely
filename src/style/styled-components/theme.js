import { colors } from './settings/colors';
import { fonts } from './settings/fonts';
import { spacing } from './settings/spacing';
import { fontSizes } from './settings/fontSizes';

export const theme = {
  fonts: { ...fonts },
  palette: { ...colors },
  spacing: { ...spacing },
  fontSizes: { ...fontSizes },
};
