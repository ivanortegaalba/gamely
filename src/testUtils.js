import 'jest-styled-components'
import React from 'react'
import { shallow } from 'enzyme'
import { ThemeProvider } from 'styled-components'



export const shallowWithTheme = (tree, theme) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext()
  return shallow(tree, { context })
}
