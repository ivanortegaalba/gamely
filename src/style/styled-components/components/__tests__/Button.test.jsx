import React from 'react';
import { Button } from '../Button';
import { shallowWithTheme } from '../../../../testUtils';
import { theme } from '../../theme';
import 'jest-styled-components';
import { px2rems } from '../../';

describe('<Button/>', () => {
  describe('Button default', () => {
    const wrapper = shallowWithTheme(<Button> Button </Button>, theme);
    it('renders without crashing', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('has default border color', () => {
      expect(wrapper).toHaveStyleRule(
        'background-color',
        theme.palette.default[1]
      );
    });
  });

  describe('Button with custom color', () => {
    const wrapper = shallowWithTheme(
      <Button palette={'primary'}>Button</Button>,
      theme
    );
    it('renders without crashing', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('has prop color', () => {
      expect(wrapper).toHaveStyleRule(
        'background-color',
        theme.palette.primary[1]
      );
    });
  });
  describe('Button with custom margin', () => {
    const component = (
      <Button
        margin={'m'}
        marginTop={'l'}
        marginBottom={'l'}
        marginLeft={'l'}
        marginRight={'l'}
      >
        Button
      </Button>
    );
    const wrapper = shallowWithTheme(component, theme);
    it('renders without crashing', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('has prop margin', () => {
      expect(wrapper).toHaveStyleRule('margin', px2rems(theme.spacing.m));
      expect(wrapper).toHaveStyleRule('margin-top', px2rems(theme.spacing.l));
      expect(wrapper).toHaveStyleRule('margin-left', px2rems(theme.spacing.l));
      expect(wrapper).toHaveStyleRule('margin-right', px2rems(theme.spacing.l));
      expect(wrapper).toHaveStyleRule(
        'margin-bottom',
        px2rems(theme.spacing.l)
      );
    });
  });
  describe('Button with custom size', () => {
    const component = <Button size={'l'}>Button</Button>;
    const wrapper = shallowWithTheme(component, theme);
    it('renders without crashing', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('has prop margin', () => {
      expect(wrapper).toHaveStyleRule('padding', '0.6875rem 1.4375rem');
    });
  });
});
