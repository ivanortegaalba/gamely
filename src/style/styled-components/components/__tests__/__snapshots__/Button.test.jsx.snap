// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> Button default renders without crashing 1`] = `
.c0 {
  background-color: #41474c;
  border: 1px solid #41474c;
  border-radius: 0.125rem;
  color: #fff;
  cursor: pointer;
  font-variant: all-small-caps;
  font-weight: 600;
  padding: 0.6875rem 1rem;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  margin: 0;
}

.c0:hover {
  background-color: #fff;
  border-color: #23272a;
  color: #23272a;
}

.c0:active,
.c0:focus {
  border-color: #59636b;
}

.c0[disabled],
.c0[disabled]:hover {
  background-color: #fff;
  border-color: #59636b;
  color: #59636b;
  cursor: inherit;
}

<button
  className="c0"
  size="m"
>
   Button 
</button>
`;

exports[`<Button/> Button with custom color renders without crashing 1`] = `
.c0 {
  background-color: #00a5c9;
  border: 1px solid #00a5c9;
  border-radius: 0.125rem;
  color: #fff;
  cursor: pointer;
  font-variant: all-small-caps;
  font-weight: 600;
  padding: 0.6875rem 1rem;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  margin: 0;
}

.c0:hover {
  background-color: #fff;
  border-color: #008aab;
  color: #008aab;
}

.c0:active,
.c0:focus {
  border-color: #43caea;
}

.c0[disabled],
.c0[disabled]:hover {
  background-color: #fff;
  border-color: #43caea;
  color: #43caea;
  cursor: inherit;
}

<button
  className="c0"
  size="m"
>
  Button
</button>
`;

exports[`<Button/> Button with custom margin renders without crashing 1`] = `
.c0 {
  background-color: #41474c;
  border: 1px solid #41474c;
  border-radius: 0.125rem;
  color: #fff;
  cursor: pointer;
  font-variant: all-small-caps;
  font-weight: 600;
  padding: 0.6875rem 1rem;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  margin-left: 1.4375rem;
  margin-right: 1.4375rem;
  margin-top: 1.4375rem;
  margin-bottom: 1.4375rem;
  margin: 1rem;
}

.c0:hover {
  background-color: #fff;
  border-color: #23272a;
  color: #23272a;
}

.c0:active,
.c0:focus {
  border-color: #59636b;
}

.c0[disabled],
.c0[disabled]:hover {
  background-color: #fff;
  border-color: #59636b;
  color: #59636b;
  cursor: inherit;
}

<button
  className="c0"
  size="m"
>
  Button
</button>
`;

exports[`<Button/> Button with custom size renders without crashing 1`] = `
.c0 {
  background-color: #41474c;
  border: 1px solid #41474c;
  border-radius: 0.125rem;
  color: #fff;
  cursor: pointer;
  font-variant: all-small-caps;
  font-weight: 600;
  padding: 0.6875rem 1.4375rem;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  margin: 0;
}

.c0:hover {
  background-color: #fff;
  border-color: #23272a;
  color: #23272a;
}

.c0:active,
.c0:focus {
  border-color: #59636b;
}

.c0[disabled],
.c0[disabled]:hover {
  background-color: #fff;
  border-color: #59636b;
  color: #59636b;
  cursor: inherit;
}

<button
  className="c0"
  size="l"
>
  Button
</button>
`;
